rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,468 @@
+network connectivity issues).
+
+Since `get` calls may be ongoing during corruption, it is possible that the
+callback will be called continuously and by multiple threads. Therefore, any
+user-defined callback must be thread-safe. Additionally, the callback should
+avoid slow or blocking operations to ensure the database remains available if
+corruption is encountered.
+
+The caller can choose to abort the application if a corruption error is
+returned by `get`. Using the details included in the corruption error, the
+application can provide operators with detailed information on the corruption,
+including suggestions on how to fix it.
+
+#### Handling Manifest Corruption during Open
+If corruption is detected in a manifest when calling `open_with_opts()`, the
+callback `DBOptions.on_corruption` is invoked. Regardless of the return value
+of the callback, `open_with_opts()` will return an error, and the database open
+operation will be aborted. A database without a valid manifest cannot operate.
+
+#### Handling SST Corruption in Cache Files
+If SlateDB detects corruption in a cache file, it should invalidate the cache
+file and remove the corrupt file. SlateDB will not invoke `DBOptions.on_corruption`
+as there is no action for the operator to take.
+
+
+#### Remaining Available during Corruption
+The rationale for keeping the database available during corruption is to
+provide options to operators and developers building complex systems on top of
+SlateDB.
+
+For instance:
+- If corruption is detected during MemTable flush, it might be desirable for
+  the developer to flush any outstanding buffers to the WAL before closing the
+  database for repair, thereby avoiding data loss.
+- If corruption is detected in the leveled SSTs, it might be desirable for the
+  developer to read uncorrupted keys from the existing SSTs into a new database
+  before fetching the corrupted or missing keys from a replica. This approach
+  reduces the time to recovery.
+- If corruption is detected, the operator may wish to remain available for
+  reads while corruption repair is taking place.
+
+However, leaving the database available for compaction and MemTable flush is a
+dangerous edge case that future patches to SlateDB might not account for.
+Therefore, this aspect should be thoroughly tested to avoid regression.
+
+Operators implementing a logical system on top of SlateDB should utilize the
+`DBOptions.on_corruption` callback to notify other SlateDB read instances of
+the corruption and take appropriate action. Operators may also use the
+`DBOptions.on_corruption` callback to signal the calling system to perform a
+safe database shutdown.
+
+#### Corruption Inspection and Repair
+To support the repair of SSTs, I propose introducing two functions:
+`sst_inspect()` and `sst_repair()`. These functions will provide developers with
+tools to inspect the current contents of a specific SST, and repair a specific
```

Comment:
how would the repair work? I suppose we could just skip corrupt data blocks. If the index or filter blocks are corrupt then we can rebuild those.